### DISCUSSION TITLE
Postgres jsonb column property has runtime type boolean

### DISCUSSION THREAD

**Discussion by pdyck** [mikro-orm/mikro-orm#5148] [Q&A] [answered]
After upgrading to v6 I can't assign anything to a particular json property. The property is persisted in Postgres as a jsonb column and can hold either a boolean, a string or a string array. Here is an example entity:

```ts
enum ContentType {
  Boolean = 'boolean',
  String = 'string',
  List = 'list',
}

type Value<Type extends ContentType> = Type extends ContentType.Boolean
  ? boolean
  : Type extends ContentType.String
    ? string
    : Type extends ContentType.List
      ? string[]
      : never

@Entity()
export class Content<Type extends ContentType> extends BaseEntity {
  @Enum({ items: () => ContentType, nullable: false })
  type: Type

  @Property({ type: 'json', nullable: true })
  value: Value<Type>
}
```

The entity metadata looks like this:

```ts
{
    columnTypes: ['jsonb'],
    customType: JsonType {...},
    fieldNames: ['value'],
    getter: false,
    hasConvertToDatabaseValueSQL: false,
    hasConvertToJSValueSQL: false,
    kind: 'scalar',
    name: 'value',
    nullable: true,
    runtimeType: 'boolean',
    setter: false,
    type: 'JsonType',
    unsigned: undefined
}
```

I did not expect `runtimeType` to be `'boolean'` but `'object'` or something more flexible. I encountered this while trying to assign a string to `Content.value` which already contained the string `'Hello world'`.

```
ValidationError: Trying to set Content.value of type 'boolean' to '<p><strong><em>Hello</em></strong>This is a change</p>' of type 'string'
```

Is this expected behavior? Is there anything I can do to configure the property so that it is less strict about validation?

**Comment by B4nan** [comment]
Can you provide a complete repro? Do you use ts-morph? It indeed shouldn't be a boolean, not sure where that could be coming from, I'd guess it's from your generic type which resolves to it.

As a workaround, you could use discovery hooks to modify the value.

https://mikro-orm.io/docs/configuration#onmetadata-hook

**Reply by pdyck** [reply]
I am using `reflect-metadata`. I will try to push a working example!

Modifying the metadata after discovery might be a nice temporary workaround, thanks.

**Reply by pdyck** [reply, marked as answer]
I'm not able to reproduce with unit tests in the `mikro-orm` repository. We are also using NestJS which comes with `reflect-metadata@0.1.x` which might be a problem. Refactoring `Value<ContentType>` to a type map resolved the issue for me:

```ts
type ValueTypeMap = {
  [ContentType.Boolean]: boolean
  [ContentType.String]: string
  [ContentType.List]: string[]
}

type Value<Type extends ContentType> = ValueTypeMap[Type]
```
